Fix incorrect result from Q.real() for Pow(0, negative exponent)

Muskan244 opened the pull request:
<!-- Your title above should be a short description of what
was changed. Do not include the issue number in the title. -->

#### References to other Issues or PRs
<!-- If this pull request fixes an issue, write "Fixes #NNNN" in that exact
format, e.g. "Fixes #1234" (see
https://tinyurl.com/auto-closing for more information). Also, please
write a comment on that issue linking back to this pull request once it is
open. -->

Fixes #28150


#### Brief description of what is fixed or changed

Fixes incorrect behavior in the assumptions system where `ask(Q.real(Pow(0, negative exponent)))` would return `True` even though such expressions are undefined due to division by zero.

An explicit check was added to `_RealPredicate_number()` to correctly identify these cases as not real.


#### Other comments

Example:

```python
from sympy import Pow, Q, ask

ask(Q.real(Pow(0, -1, evaluate=False)))  # returns False now
```

Test cases have been added to `sympy/assumptions/tests/test_query.py` to verify this behavior.

#### Release Notes

<!-- Write the release notes for this release below between the BEGIN and END
statements. The basic format is a bulleted list with the name of the subpackage
and the release note for this PR. For example:

* solvers
  * Added a new solver for logarithmic equations.

* functions
  * Fixed a bug with log of integers. Formerly, `log(-x)` incorrectly gave `-log(x)`.

* physics.units
  * Corrected a semantical error in the conversion between volt and statvolt which
    reported the volt as being larger than the statvolt.

or if no release note(s) should be included use:

NO ENTRY

See https://github.com/sympy/sympy/wiki/Writing-Release-Notes for more
information on how to write release notes. The bot will check your release
notes automatically to see if they are formatted correctly. -->

<!-- BEGIN RELEASE NOTES -->

* assumptions
  * Correctly handle division by zero in `Q.real()` for expressions like `Pow(0, negative exponent)`.

<!-- END RELEASE NOTES -->

Muskan244 committed:
assumptions: correctly handle division by zero in _RealPredicate_number

Previously, `ask(Q.real(Pow(0, -1, evaluate=False)))` incorrectly returned True,
because the assumptions system only structurally analyzed the imaginary part of
the expression, without checking for undefined domain cases like division by zero.

This commit adds an explicit check in `_RealPredicate_number()` to return False
when the expression is Pow(0, n) with negative n, which corresponds to division
by zero and is undefined over the real numbers.

Fixes: #28150

Example:

    >>> from sympy import Pow, Q, ask
    >>> ask(Q.real(Pow(0, -1, evaluate=False)))
    False

Test cases have been added to sympy/assumptions/tests/test_query.py to verify
the correct behavior for these cases.

Muskan244 committed:
author: add Muskan Raghav to .mailmap

Muskan244 committed:
author: add Muskan Raghav to .mailmap

krishnavbajoria02 commented on sympy/assumptions/handlers/sets.py:
```diff
@@ -230,6 +230,35 @@ def _(expr, assumptions):
+    if isinstance(expr, Pow):
+        base, exp = expr.args
+        # Check if it's Pow(0, negative number)
+        if base == 0 and exp.is_number and exp.evalf() < 0:
```
I don't think we need a check for ```exp.is_number``` as this handler is only called when the ```expr``` is a number.

krishnavbajoria02 commented on sympy/assumptions/handlers/sets.py:
```diff
@@ -230,6 +230,35 @@ def _(expr, assumptions):
+
+    """
+    if isinstance(expr, Pow):
+        base, exp = expr.args
```
We can just use ```expr.base``` and ```expr.exp``` here and we don't need to initialize two new variables.

TiloRC commented on sympy/assumptions/handlers/sets.py:
```diff
@@ -230,6 +230,35 @@ def _(expr, assumptions):
+    if isinstance(expr, Pow):
+        base, exp = expr.args
+        # Check if it's Pow(0, negative number)
+        if base == 0 and exp.is_number and exp.evalf() < 0:
```
We should avoid using `evalf` because it's expensive.

Muskan244 replied:
Thank you! I will replace evalf() with is_negative to avoid numeric evaluation.

Muskan244 commented on sympy/assumptions/handlers/sets.py:
```diff
@@ -230,6 +230,35 @@ def _(expr, assumptions):
 # RealPredicate
 
 def _RealPredicate_number(expr, assumptions):
+    """
+    Evaluate whether a given numeric expression is real.
+
+    This function is used by the assumptions system to determine whether
+    a numeric expression can be considered real. It first checks for
+    known undefined cases such as division by zero in Pow(0, negative number),
+    and then applies structural checks based on the imaginary part.
+
+    Examples
+    ========
+
+    >>> from sympy import Pow, Q, ask
+
+    >>> ask(Q.real(Pow(0, -1, evaluate=False)))
+    False
+
+    >>> ask(Q.real(Pow(0, 2, evaluate=False)))
+    True
+
+    >>> ask(Q.real(Pow(0, 0, evaluate=False)))
+    True
+
+    """
```
Thank you! I will remove the docstring as suggested.

Muskan244 commented on sympy/assumptions/tests/test_query.py:
```diff
@@ -2569,3 +2569,6 @@ def test_issue_25221():
     nan = S.NaN
     assert ask(Q.negative(nan)) is None
+
+def test_real_pow_zero_negative_exponent():
+    assert ask(Q.real(Pow(0, -1, evaluate=False))) == False
```
Thank you! I will add the issue number to the test.

Muskan244 committed:
assumptions: simplify check in _RealPredicate_number and add regression test for issue #28150

krishnavbajoria02 commented on sympy/assumptions/tests/test_query.py:
```diff
@@ -2569,3 +2569,6 @@ def test_issue_25221():
     assert ask(Q.negative(nan)) is None
+
+def test_issue_28150():
+    assert ask(Q.real(Pow(0, -1, evaluate=False))) == False
```
You should add a comment with the issue link as mentioned in the link suggested by Tilo.

Muskan244 replied:
Thank you! I will add the comment with the issue link and push the changes shortly.

Muskan244 committed:
tests: add issue reference and improve comparison in test_issue_28150

TiloRC commented on sympy/assumptions/tests/test_query.py:
```diff
@@ -2569,3 +2569,7 @@ def test_issue_25221():
     assert ask(Q.negative(nan)) is None
+
+def test_issue_28150():
+    # https://github.com/sympy/sympy/issues/28150
```
You don't need the comment since the issue number is in the test name

krishnavbajoria02 replied:
Ohh sorry, didn’t notice that. Should probably remove that comment then.

Muskan244 replied:
Got it! Since I had already added the comment as per the earlier suggestion, I’ll leave it for now — but happy to remove it if you prefer.

Muskan244 replied:
> Ohh sorry, didn’t notice that. Should probably remove that comment then.

Thank you! I'll remove the comment.

krishnavbajoria02 replied:
Yes you should remove it. It’s not needed since the issue number is mentioned in the function definition.

TiloRC commented on sympy/assumptions/handlers/sets.py:
```diff
@@ -230,6 +230,11 @@ def _(expr, assumptions):
 
 def _RealPredicate_number(expr, assumptions):
+    if isinstance(expr, Pow):
+        # Check if it's Pow(0, negative number)
```
I don't think this comment is needed

Muskan244 replied:
Thank you! I'll remove the comment.

TiloRC commented:
I found a case that is still wrongly giving False:
```python
>>> from sympy import Pow, Q, ask, Mul
>>> zero = Mul(0, 1, evaluate=False)
>>> ask(Q.real(Pow(zero, -1, evaluate=False)))  # returns False now
True
```

Muskan244 committed:
assumptions: handle unevaluated zero base in Pow and clean up comments

Extended _RealPredicate_number to correctly detect Pow with unevaluated zero base (e.g. Mul(0, 1, evaluate=False)) and negative exponent as not real. Also removed unnecessary comments from test code as discussed in review.

krishnavbajoria02 commented on sympy/assumptions/handlers/sets.py:
```diff
@@ -230,6 +230,10 @@ def _(expr, assumptions):
 
 def _RealPredicate_number(expr, assumptions):
+    if isinstance(expr, Pow):
+        if expr.base.equals(0) and expr.exp.is_negative:
```
I have not usually seen ```equals()``` be used in assumptions a lot. I would prefer here using ```ask(Q.eq(expr.base,0), assumptions)``` .

Muskan244 replied:
Thanks for clarifying! I didn’t realize that .equals() internally calls simplify() and can affect performance. I’ll switch to ask(Q.eq(expr.base,0), assumptions) for better efficiency.

Muskan244 committed:
assumptions: improve performance by replacing equals() with ask(Q.eq(...))

The use of .equals() in assumptions handlers is avoided as it internally calls simplify(), which can be expensive. Updated _RealPredicate_number() to use ask(Q.eq(expr.base, 0), assumptions) for better efficiency.

TiloRC commented:
The approach you're currently using is not going to work. The problem of determining whether an expression is equal to zero is undecidable, so we will never be able to be sure the base is not equal to zero. Also, we should not be using the recursive handlers here as this method should be mostly self contained.  We need to change the way this part of the code works:

```python
# let as_real_imag() work first since the expression may
# be simpler to evaluate
i = expr.as_real_imag()[1].evalf(2)
if i._prec != 1:
    return not i
# allow None to be returned if we couldn't show for sure
# that i was 0
```

Muskan244 commented:
Thanks for the detailed explanation! I’ll work on updating the function to use direct pattern checks and keep it self-contained without relying on ask() or evalf(). Will push the revised version shortly!

TiloRC commented:
I think evalf() is probably fine. However, we should be using [im](https://docs.sympy.org/latest/modules/functions/elementary.html#sympy.functions.elementary.complexes.im) instead of  as_real_imag I think as the latter performs expansions that might be expensive.

Muskan244 commented:
Thanks! That makes sense — I'll update it to use expr.im instead of as_real_imag() for better performance.

Muskan244 commented:
I tried using .im but ran into AttributeErrors for objects like Pow, Add, and Mul.
Would it be okay to try using .im first, and then fallback to .as_real_imag()[1] (with a try/except) if .im isn't available? This way we avoid recursion and still cover most cases.
Let me know if this approach makes sense or if there’s a better way to handle this.

Muskan244 committed:
fix(assumptions): avoid undecidable zero checks and recursive handlers in _RealPredicate_number

Muskan244 commented:
Thanks for the feedback! I've pushed the new changes — replaced ask(Q.eq(expr.base,0), assumptions) with .is_zero, and used im(expr) instead of as_real_imag() to avoid expensive expansions as suggested. Let me know if there’s anything else I should refine!

Muskan244 commented:
Hi! I’ve pushed the requested changes, and it looks like all checks have passed. Just checking in to see if there's anything more I should do. Thanks again for guiding me through this!

TiloRC commented:
There are still a lot of cases where this currently gives the wrong result:

```python
>>> from sympy import ask, Q, Pow, S
>>> ask(Q.real(Pow(S.Infinity, -1, evaluate=False))) is False
False
>>> ask(Q.real(Pow(S.Infinity, -1, evaluate=False)))
True
>>> ask(Q.real(Pow(S.ComplexInfinity, -1, evaluate=False)))
True
>>> ask(Q.real(Pow(S.NaN, -1, evaluate=False)))
False
>>> ask(Q.real(Pow(Pow(0,-1, evaluate=False), -1, evaluate=False))) # raises exception when it shouldn't
```

TiloRC commented:
> The approach you're currently using is not going to work. The problem of determining whether an expression is equal to zero is undecidable, so we will never be able to be sure the base is not equal to zero.

This still applies

TiloRC commented on sympy/assumptions/handlers/sets.py:
```diff
@@ -230,9 +230,13 @@ def _(expr, assumptions):
 
 def _RealPredicate_number(expr, assumptions):
+    if isinstance(expr, Pow):
+        if expr.base.is_zero and expr.exp.is_negative:
```
I'm not sure the `_number` functions are actually a good idea in the first place. They all seem to work by creating new, simpler expressions, which seems very expensive. I would think the assumption system as a whole should probably avoid making new expressions like this entirely.

Muskan244 commented:
> There are still a lot of cases where this currently gives the wrong result:
> 
> ```python
> >>> from sympy import ask, Q, Pow, S
> >>> ask(Q.real(Pow(S.Infinity, -1, evaluate=False)))
> True
> >>> ask(Q.real(Pow(S.Infinity, -1, evaluate=False)))
> True
> >>> ask(Q.real(Pow(S.ComplexInfinity, -1, evaluate=False)))
> True
> >>> ask(Q.real(Pow(S.NaN, -1, evaluate=False)))
> False
> >>> ask(Q.real(Pow(Pow(0,-1, evaluate=False), -1, evaluate=False))) # raises exception when it shouldn't
> ```

I noticed that in SymPy, Pow(S.Infinity, -1) and Pow(S.ComplexInfinity, -1) actually gives 0. According to the [SymPy documentation on Infinity and ComplexInfinity](https://docs.sympy.org/latest/modules/core.html#sympy.core.numbers.Infinity):
> ```python
> >>> 42/oo
> 0
> >>> 42/zoo
> 0

Since 0 is real and finite, ask(Q.real(Pow(S.Infinity, -1, evaluate=False))) and ask(Q.real(Pow(S.ComplexInfinity, -1, evaluate=False))) returns True.

Also, Pow(S.NaN, -1) gives nan. So, if ask(Q.real(Pow(S.NaN, -1, evaluate=False))) gives False, is that correct, or should it return None instead?

I just wanted to check if this matches what’s expected, or if there’s a different interpretation I should be following here? If I’m missing something, I’d really appreciate any clarification!

Muskan244 committed:
fix(assumptions): return None for NaN-related queries and avoid exceptions for special values like zoo

Muskan244 committed:
fix(power): remove duplicate method definitions to resolve F811 ruff errors

Muskan244 committed:
remove redefinitions in sympy/core/power.py

Muskan244 committed:
remove redefinitions in sympy/core/power.py

TiloRC commented:
You shouldn't be modifying anything from the core directory. This bug is related to the new, not old assumptions.

Muskan244 commented:
Thanks for pointing it out! I’ll revert my changes to `sympy/core/power.py` and move all the special-case logic into the new assumptions handlers instead, so the core isn’t modified.

Muskan244 committed:
Fix Q.real handler to safely handle undefined and nested Pow cases without core changes.
